src/backend/distributed/operations/shard_split.c

```diff
@@ -1145,3 +1221,650 @@ TryDropSplitShardsOnFailure(HTAB *mapOfShardToPlacementCreatedByWorkflow)
+		List *replicationSlotInfoList = ExecuteSplitShardReplicationSetupUDF(
+			sourceShardToCopyNode,
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			workersForPlacementList);
+
+		/*
+		 * Subscriber flow starts from here.
+		 * Populate 'ShardSplitSubscriberMetadata' for subscription management.
+		 */
+		List *shardSplitSubscribersMetadataList =
+			PopulateShardSplitSubscriptionsMetadataList(
+				shardSplitHashMapForPublication, replicationSlotInfoList);
+
+		/* Create connections to the target nodes. TODO: can be refactored */
+		List *targetNodeConnectionList = CreateTargetNodeConnectionsForShardSplit(
+			shardSplitSubscribersMetadataList,
+			connectionFlags,
+			superUser, databaseName);
+
+		/* 7) Create copies of template replication slot */
+		char *templateSlotName = ShardSplitTemplateReplicationSlotName(
+			shardIntervalToSplit->shardId);
+		CreateReplicationSlots(sourceConnection, templateSlotName,
+							   shardSplitSubscribersMetadataList);
+
+		/* 8) Create subscriptions on target nodes */
+		CreateShardSplitSubscriptions(targetNodeConnectionList,
+									  shardSplitSubscribersMetadataList,
+									  sourceShardToCopyNode,
+									  superUser,
+									  databaseName);
+
+		/* 9) Wait for subscriptions to be ready */
+		WaitForShardSplitRelationSubscriptionsBecomeReady(
+			shardSplitSubscribersMetadataList);
+
+		/* 10) Wait for subscribers to catchup till source LSN */
+		XLogRecPtr sourcePosition = GetRemoteLogPosition(sourceConnection);
+		WaitForShardSplitRelationSubscriptionsToBeCaughtUp(sourcePosition,
+														   shardSplitSubscribersMetadataList);
+
+		/* 11) Create Auxilary structures */
+		CreateAuxiliaryStructuresForShardGroup(shardGroupSplitIntervalListList,
+											   workersForPlacementList);
+
+		/* 12) Wait for subscribers to catchup till source LSN */
+		sourcePosition = GetRemoteLogPosition(sourceConnection);
+		WaitForShardSplitRelationSubscriptionsToBeCaughtUp(sourcePosition,
+														   shardSplitSubscribersMetadataList);
+
+		/* 13) Block writes on source shards */
+		BlockWritesToShardList(sourceColocatedShardIntervalList);
+
+		/* 14) Wait for subscribers to catchup till source LSN */
+		sourcePosition = GetRemoteLogPosition(sourceConnection);
+		WaitForShardSplitRelationSubscriptionsToBeCaughtUp(sourcePosition,
+														   shardSplitSubscribersMetadataList);
+
+		/* 15) Drop Subscribers */
+		DropShardSplitSubsriptions(shardSplitSubscribersMetadataList);
+
+		/* 16) Drop Publications */
+		DropShardSplitPublications(sourceConnection, shardSplitHashMapForPublication);
+
+		/* 17) TODO(saawasek): Try dropping replication slots explicitly */
+
+		/*
+		 * 18) Drop old shards and delete related metadata. Have to do that before
+		 * creating the new shard metadata, because there's cross-checks
+		 * preventing inconsistent metadata (like overlapping shards).
+		 */
+		DropShardList(sourceColocatedShardIntervalList);
```

Comment:
From this point, would be nice to refactor with blocking split to avoid duplicates.

Reply:
I had given this thought.  But felt that since the nonblocking split has lot of steps placing individual helpers would be good. Will give a thought again.